ampd/src/starknet/json_rpc.rs

```diff
@@ -0,0 +1,689 @@
+
+use std::str::FromStr;
+
+use async_trait::async_trait;
+use mockall::automock;
+use starknet_core::types::{
+    ExecutionResult, FieldElement, FromStrError, MaybePendingTransactionReceipt, TransactionReceipt,
+};
+use starknet_providers::jsonrpc::JsonRpcTransport;
+use starknet_providers::{JsonRpcClient, Provider, ProviderError};
+use thiserror::Error;
+
+use crate::starknet::events::contract_call::ContractCallEvent;
+
+#[derive(Debug, Error)]
+pub enum StarknetClientError {
+    #[error(transparent)]
+    UrlParseError(#[from] url::ParseError),
+    #[error(transparent)]
+    JsonDeserializeError(#[from] serde_json::Error),
+    #[error("Failed to fetch tx receipt: {0}")]
+    FetchingReceipt(#[from] ProviderError),
+    #[error("Failed to create field element from string: {0}")]
+    FeltFromString(#[from] FromStrError),
+    #[error("Tx not successful")]
+    UnsuccessfulTx,
+}
+
+/// Implementor of verification method(s) for given network using JSON RPC
+/// client.
+pub struct Client<T>
+where
+    T: JsonRpcTransport + Send + Sync + 'static,
+{
+    client: JsonRpcClient<T>,
+}
+
+impl<T> Client<T>
+where
+    T: JsonRpcTransport + Send + Sync + 'static,
+{
+    /// Constructor.
+    /// Expects URL of any JSON RPC entry point of Starknet, which you can find
+    /// as constants in the `networks.rs` module
+    pub fn new(transport: T) -> Result<Self, StarknetClientError> {
+        Ok(Client {
+            client: JsonRpcClient::new(transport),
+        })
+    }
+}
+
+#[automock]
+#[async_trait]
+pub trait StarknetClient<T>
+where
+    T: JsonRpcTransport + Send + Sync + 'static,
+{
+    async fn get_event_by_hash(
+        &self,
+        tx_hash: &str,
+    ) -> Result<Option<(String, ContractCallEvent)>, StarknetClientError>;
+}
+
+#[async_trait]
+impl<T> StarknetClient<T> for Client<T>
+where
+    T: JsonRpcTransport + Send + Sync + 'static,
+{
+    /// Using given transaction hash, tries to fetch it from given
+    /// `starknet_url`. Returns error if request fails, `false` if internal
+    /// error returned by querry and `true` if transaction found
+    async fn get_event_by_hash(
+        &self,
+        tx_hash: &str,
+    ) -> Result<Option<(String, ContractCallEvent)>, StarknetClientError> {
+        let tx_hash_felt = FieldElement::from_str(tx_hash)?;
+
+        // TODO: Check ACCEPTED ON L1 times and decide if we should use it
+        //
+        // Finality status is always at least ACCEPTED_ON_L2 and this is what we're
+        // looking for, because ACCEPTED_ON_L1 (Ethereum) will take a very long time.
+        let receipt_type = self.client.get_transaction_receipt(tx_hash_felt).await?;
+
+        if *receipt_type.execution_result() != ExecutionResult::Succeeded {
+            return Err(StarknetClientError::UnsuccessfulTx);
+        }
+
+        let event: Option<(String, ContractCallEvent)> = match receipt_type {
+            // TODO: There is also a PendingReceipt type. Should we handle it?
```

Comment:
Not sure what the time difference is between PendingReceipt and a normal Receipt.
I'm also not sure if we can accidentally skip processing a message, in case axelar receives a message and emits an event before the starknet tx recipt has moved from pending to normal state.

I think it's not possible, but I don't know the time a tx goes from pending to not pending, in order to say for sure.

Reply:
Create [#90](https://github.com/eigerco/giza-axelar-starknet/issues/90) to tackle it